### DISCUSSION TITLE
Use GitLab CODEOWNERS approvals with Renovate automerge

### DISCUSSION THREAD

**Discussion by JuJup** [renovatebot/renovate#32907] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab

### Please tell us more about your question or problem

**Problem**
GitLab offers the option to enable [Code Owner approval rules](https://docs.gitlab.com/ee/user/project/codeowners/), which require an approval before MRs can be merged.
This leads to the problem, that Renovate cannot automerge the created MRs, before a Code Owner gives it's approval. This might be a wanted behaviour on some updates, but it conflicts with the goal to have [less noise by using automerge](https://docs.renovatebot.com/noise-reduction/#automerging).
The [gitlabIgnoreApprovals](https://docs.renovatebot.com/configuration-options/#gitlabignoreapprovals) option does not solve the problem, as Code Owner rules can't be changed on a MR level.

There are already some Issues and Discussion about it (https://github.com/renovatebot/renovate/discussions/25905, https://github.com/renovatebot/renovate/discussions/18328, https://github.com/renovatebot/renovate/discussions/23052, https://github.com/renovatebot/renovate/discussions/25224, https://github.com/renovatebot/app-support/issues/43), but all the workarounds feel hacky somehow and I want to share my research results - maybe I forgot something or you have better ideas? And also, maybe the documentation can be extended to have some advices for users with the same problem. Especially on the `gitlabIgnoreApprovals` option in the docs there should be a note, that Code Owner Rules can't be changed in a MR.

**Possible Workarounds**
1. use [autoApprove](https://docs.renovatebot.com/configuration-options/#autoapprove) option
    -  you cannot use the option `Prevent approval by author`
    - the Bot User has to be added as Code Owner to all files which can be automerged
    - only done on MR creation - [on rebase approvals will be resetted automatically and not added again](https://github.com/renovatebot/renovate/discussions/31833)
1. use another Bot to approve changes (see https://github.com/renovatebot/renovate/discussions/18328#discussioncomment-3873820)
    - seperation of concerns
    - you can use the option `Prevent approval by author`
    - the Bot User still has to be added as Code Owner to all files which can be automerged
    - the [Renovate Approve Bot](https://github.com/apps/renovate-approve) is only for GitHub
    - overhead because you have to run additional jobs for the approvals
2. use [automergeType=branch](https://docs.renovatebot.com/configuration-options/#automergetype)
    - Renovate needs rights to push directly into main branch
    - we can't use benefit from [merge request pipelines](https://docs.gitlab.com/ee/ci/pipelines/merge_request_pipelines.html)
4. remove the files where we want to use automerge from Code Owners
    - Code Owner files don't have an exclude syntax, so it's gonna be hard to maintain
    - many files (.gitlab-ci.yml, Dockerfile, etc.) have way more content than dependencies, so you'd loose the benefit of ensuring quality with Code Owners
5. don't use Code Owner approval rules
    - we'd loose the great benefit of ensuring quality by defining specific approver for certian files and directories
6. live with it
    - Renovate will enable auto-merge on the MR, which means the MR will be merged directly after the Code Owner approves (when the other checks succeeded)
    - we'd still get unnecessary noise

Any thoughts on this? Is there a recommended way by Renovate?

**Comment by rarkins** [comment]
Making Renovate a codeowner of all seems like the most preferred (other than using a separate type of bot for approvals).

**Comment by carlos-lehmann** [comment]
Hey @rarkins @JuJup 

Hijacking this thread because we are running into a similar use case here with GitLab and self hosted Renovate:

- we have automerge configured:

```json5
      "schedule": [
        "at any time"
      ],
      "platformAutomerge": false,
      "automerge": true,
      "automergeType": "pr",
      "automergeStrategy": "squash",
      "automergeSchedule": [
        "after 12pm on Thursday"
      ],
```

- we use CODEOWNERS approval (renovate is in CODEOWNERS)

```
[Maintainers]  @TEAM-A @TEAM-B @renovate
*
```

- we use `Prevent approval by author`
- we use `gitLabIgnoreApprovals: true`

We get a 405 in the debug output:
```json
        "message": "Response code 405 (Method Not Allowed)",
         "stack": "HTTPError: Response code 405 (Method Not Allowed)\n    at Request.<anonymous> (/usr/local/renovate/node_modules/.pnpm/got@11.8.6/node_modules/got/dist/source/as-promise/index.js:118:42)\n    at processTicksAndRejections (node:internal/process/task_queues:105:5)",
```

Please correct me if I'm wrong but reading through [gitLabIgnoreApprovals](https://docs.renovatebot.com/configuration-options/#gitlabignoreapprovals) it says: `Under the hood, it creates a MR-level approval rule where approvals_required is set to 0`. So it can only reset `approvals_required` and does not manage the flag `Prevent approval by author`. Is that correct?

Is there another flag in renovate which I did not see that I could use? Is it on the roadmap to be supported? Maybe there's a catch in Gitlab that you can't manage it I did not understand yet? 

Additionally, feature request: We would really appreciate that if automerge has an error, that we would get an information that automerge did not succeed somewhere (dependency dashboard?). 

Looking forward to your answer,

Thanks a lot!
Carlos

**Comment by carlos-lehmann** [comment]
For anyone reading this and having similar issues, we now went with the following configuration.

In Gitlab:

1. Remove `Prevent approval by author` and replace it with `Minimum required approvals` = `2`
2. Remove `CODEOWNERS approval`

In Renovate Config:

3. Use `gitLabIgnoreApprovals` = `true` 

This allows renovate at MR creation to override the minimum required approvals to `0`. It's not ideal but with the limited knowledge I have, whether this is something that is missing in Renovate and/or Gitlab, this is a working solution. This allows us to use Renovate with automerge AND having at least some level of security with 4eye.

**Reply by declension** [reply]
This inspired us to have another go very similar to this, and it worked - thanks!

Our take-aways:
* We managed to keep `CODEOWNERS` working, by creating a [service account](https://docs.gitlab.com/user/profile/service_accounts/) with a nice username (note @carlos-lehmann that `@renovate` is probably not correct for you either), and including this in each repo we wanted. Not the best but better than ditching CODEOWNERs for us.
* This route has definitely got sharp edges (the ability for humans to override approval settings is _not_ cool, and two-approvals-but-only-one-really is kinda weird).
* Longer term we might look at branch-based Renovation ([`automergeType=branch`](https://docs.renovatebot.com/configuration-options/#automergetype)) instead, as the security seems a bit better, even it'd mean a lot of Gitlab `rules`-changes for us.
